### PR TITLE
Start keeping a stack of namespaces instead of namespace names

### PR DESCRIPTION
**Pull request by paracycle**
### Motivation

The current method collection was assigning the wrong owner for methods defined after a namespace was entered and then exited, since the `@current_owner` was not being "popped" at all.

For example:
```ruby
module Foo
  module Bar
  end

  def foo
  end
end
```
would end up assigning `Foo::Bar` as the owner of the method `foo`, when it should actually be `Foo`.

### Implementation

This allows us to use the same stack for both the constant and method collection. For constants, keeping the current namespace on the stack allows us to refer to the current namespace name instantly, without having to build it from the stack entries. For methods, we need to keep the current namespace on the stack so that we can properly find the owner, i.e. the namespace that the method is defined in.

### Automated Tests

Added a few automated tests to test the initial symptom, plus a few edge cases.

### CODE REVIEW

File: lib/ruby_indexer/test/constant_test.rb [hunk 1]
```diff
@@ -305,6 +307,25 @@ module A
...
+      assert_instance_of(Entry::UnresolvedAlias, unresolve_entry)
+      assert_nil(unresolve_entry.namespace)
+      assert_equal("O::P", unresolve_entry.target)
+
+      resolved_entry = @index.resolve("N", []).first
+      assert_instance_of(Entry::Alias, resolved_entry)
+      assert_equal("O::P", resolved_entry.target)
+
+      # When we are resolving A::I::J, we invoke `resolve("K::L", ["A"])`, which recursively resolves A::K::L too.
+      # Therefore, both A::I::J and A::K::L point to A::M by the end of the previous resolve invocation
```
Comment (st0012) [inline lib/ruby_indexer/test/constant_test.rb:322]: I think these comments are mistakenly copied here?

File: lib/ruby_indexer/test/constant_test.rb [hunk 2]
```diff
@@ -305,6 +307,25 @@ module A
...
+      resolved_entry = @index["O::P"].first
+      assert_instance_of(Entry::Alias, resolved_entry)
+      assert_equal("Q::R", resolved_entry.target)
+
+      constant = @index["Q::R"].first
+      assert_instance_of(Entry::Constant, constant)
```
Comment (st0012) [inline lib/ruby_indexer/test/constant_test.rb:328]: Can we have a new test case for these new assertions? It took me a while to realise they actually don't rely on other parts of the fixture.

**Review by st0012 — COMMENTED** [review]
Left a few comments on tests. Otherwise the fix looks good 👍

File: lib/ruby_indexer/lib/ruby_indexer/index.rb [hunk 3]
```diff
@@ -255,7 +255,7 @@ def resolve_method(method_name, receiver_name)
     # that doesn't exist, then we return the same UnresolvedAlias
     sig { params(entry: Entry::UnresolvedAlias).returns(T.any(Entry::Alias, Entry::UnresolvedAlias)) }
     def resolve_alias(entry)
-      target = resolve(entry.target, entry.nesting)
+      target = resolve(entry.target, Array(entry.namespace&.name))
```
Comment (vinistock) [inline lib/ruby_indexer/lib/ruby_indexer/index.rb:258]: While this might work, the idea behind the second argument is for it to be an array containing each part of the name.

I.e.:
```ruby
"Foo::Bar::Baz"
# Should be
["Foo", "Bar", "Baz"]
```

What do you think about adding a method named `nesting` in `UnresolvedAlias` that just splits the name?

```ruby
class UnresolvedAlias
  def nesting
    @namespace.name.split("::")
  end
end
```

**Review by vinistock — COMMENTED** [review]
This is great ❤️

File: lib/ruby_indexer/lib/ruby_indexer/index.rb (hunk 3 above)
Comment (paracycle) [reply lib/ruby_indexer/lib/ruby_indexer/index.rb:258]: Sure, I can make that change but I think this part of the indexer is fundamentally broken, and needs a refactor anyway. The problem is that Ruby nesting is not as simple as looking at each part of the namespace, since implicit namespaces work differently:
```ruby
module Foo
  A = 1
end

module Foo::Bar
  puts A
  # => uninitialized constant Foo::Bar::A (NameError)
end
```

For more reference see: https://www.writesoftwarewell.com/nested-modules-in-ruby/

Thus, what we really need to do is to store the "nesting" as Ruby would have and resolve against that nesting stack.

File: lib/ruby_indexer/test/constant_test.rb (hunk 2 above)
Comment (paracycle) [reply lib/ruby_indexer/test/constant_test.rb:328]: I don't understand how these are not related. This test case is testing all kinds of ways constant aliases could be defined, but it was never checking for top level constant alias definitions, which were broken.

My changes just add top-level constant alias definition examples to this method as well.

File: lib/ruby_indexer/lib/ruby_indexer/index.rb (hunk 3 above)
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/index.rb:258]: I think that's already what happens in main with the naming stack. Unless I'm missing something, I don't think it's broken.

We don't split namespaces that are defined as `module Foo::Bar`, so I believe the resolution is correct. Although we should add a test for that specific example.

The changes in this PR might require some extra refactors to make it fully work. For example, if we found an alias like this

```ruby
module First
  class Foo::Bar
    ALIAS_CONST = SOME_OTHER_THING
  end
end
```

We correctly store the nesting as `["First", "Foo::Bar"]`. When resolving `SOME_OTHER_THING`, we'd then search it in `First::Foo::Bar`, then `First` and finally on the top level.

And the features that use `resolve` do send the right nesting. For example, definition would invoke it like `@index.resolve("SOME_OTHER_THING", ["First", "Foo::Bar"])` - since we also don't split those names when locating target nodes for features.

However, with these changes we no longer have a reference to the nesting where things are found for unresolved symbols. We only have the entries with their fully qualified names in the stack, which as you pointed out is not enough to reconstruct the nesting where they were originally found.

To make the change in this PR and not break unresolved aliases, I think we need to change the entries to store their nesting instead of the fully qualified name, so that we can work back the nesting for aliases. Either that or maintain the original naming stack, but that seems more wasteful.

Comment (paracycle) [reply lib/ruby_indexer/lib/ruby_indexer/index.rb:258]: Right, that makes sense. That also means that we are missing a lot of test cases that should have broken because of my changes in this PR, exercising specifically those kinds of nesting cases.
